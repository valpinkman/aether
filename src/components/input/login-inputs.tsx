import React, { SyntheticEvent, useRef, useState } from 'react'
import styled, { keyframes } from 'styled-components'
import { User } from 'styled-icons/boxicons-solid/User'
import { DoorLock } from 'styled-icons/remix-fill/DoorLock'
import { Eye } from 'styled-icons/remix-fill/Eye'
import { EyeOff } from 'styled-icons/remix-fill/EyeOff'
import { Spinner4 } from 'styled-icons/icomoon/Spinner4'

import Box from '../box/box'
import Input from './input'


const Wrapper = styled(Box)`
  border-radius: 48px;
  background-color: ${p => p.theme.colors.background.default};
  padding: ${p => `${p.theme.space[3]}px ${p.theme.space[4]}px`};
  align-items: center;
  box-shadow: 0px 6px 12px ${p =>
    p.focused ? 'rgba(60, 90, 120, 0.2)' : 'rgba(60, 90, 120, 0.1)'};
  transition: box-shadow ease-out 300ms;
`

const ShowingBox = styled(Box)`
  cursor: pointer;
  transition: opacity ease-out 200ms;
`
ShowingBox.displayName = 'ShowingBox'

const EmailIcon = styled(User)`
  color: ${p => p.theme.colors.text.accent};
`

const DoorLockIcon = styled(DoorLock)`
  color: ${p => p.theme.colors.text.accent};
`
const EyeIcon = styled(Eye)`
  color: ${p => p.theme.colors.text.accent};
`
EyeIcon.displayName = 'EyeIcon'

const EyeOffIcon = styled(EyeOff)`
  color: ${p => p.theme.colors.text.accent};
`
EyeOffIcon.displayName = 'EyeOffIcon'

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`

const ArrowSyncIcon = styled(Spinner4)`
  color: ${p => p.theme.colors.text.accent};
  animation: ${rotate} 1s alternate-reverse infinite;
`
ArrowSyncIcon.displayName = 'ArrowSyncIcon'

const WithSpinner = ({ loading, children }: { loading: boolean; children: any }) => loading
  ? <ArrowSyncIcon size="20px" />
  : children

type InputProps = {
  onChange: (e: SyntheticEvent<HTMLInputElement>) => void;
  loading: boolean;
}

export const EmailInput = ({ onChange, loading }: InputProps) => {
  const [hovered, setHovered] = useState(false)

  return (
    <Wrapper
      title="wrapper"
      flexDirection="row"
      width="400px"
      focused={hovered}
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
    >
      <Box pr={2}>
        <WithSpinner loading={loading}>
          <EmailIcon size="20px" />
        </WithSpinner>
      </Box>
      <Input
        type="email"
        fontSize={2}
        onChange={onChange}
        bg="transparent"
        width="100%"
        maxWidth="100%"
        color="text.accent"
        placeholder="email"
      />
    </Wrapper>
  )
}

export const PasswordInput = ({ onChange, loading }: InputProps) => {
  const [hovered, setHovered] = useState(false)
  const input = useRef(null)
  const [isRevealed, setIsRevealed] = useState(false)

  return (
    <Wrapper
      title="wrapper"
      flexDirection="row"
      width="400px"
      focused={hovered}
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
    >
      <Box pr={2}>
        <WithSpinner loading={loading}>
          <DoorLockIcon size="20px" />
        </WithSpinner>
      </Box>
      <Input
        ref={input}
        type={isRevealed ? 'text' : 'password'}
        fontSize={2}
        onChange={onChange}
        bg="transparent"
        width="100%"
        maxWidth="100%"
        color="text.accent"
        placeholder="password"
      />
      <ShowingBox
        pl={1}
        onClick={() => setIsRevealed(!isRevealed)}
        opacity={hovered ? 1 : 0}
      >
        {isRevealed ? <EyeIcon size="16px" /> : <EyeOffIcon size="16px" />}
      </ShowingBox>
    </Wrapper>
  )
}

export default Input
