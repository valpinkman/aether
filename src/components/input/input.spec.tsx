import React from 'react'
import { shallow } from 'enzyme'

import Input from './input'
import { EmailInput, PasswordInput } from './login-inputs'

const onChangeCallback = jest.fn()

describe('components', () => {
  describe('<Input />', () => {
    it('should render', () => {
      const wrapper = shallow(<Input />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('<EmailInput />', () => {
    it('should render', () => {
      const wrapper = shallow(<EmailInput onChange={onChangeCallback} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('<PasswordInput />', () => {
    it('should render', () => {
      const wrapper = shallow(<PasswordInput onChange={onChangeCallback} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should switch icon and input type', () => {
      const wrapper = shallow(<PasswordInput onChange={onChangeCallback} />)


      wrapper.simulate('mouseenter')
      wrapper.update()

      expect(wrapper.find(Input).prop('type')).toBe('password')
      expect(wrapper.find('EyeOffIcon')).toHaveLength(1)
      expect(wrapper.find('EyeIcon')).toHaveLength(0)

      wrapper.find('ShowingBox').simulate('click')
      wrapper.update()

      expect(wrapper.find(Input).prop('type')).toBe('text')
      expect(wrapper.find('EyeOffIcon')).toHaveLength(0)
      expect(wrapper.find('EyeIcon')).toHaveLength(1)
    })

    it('should trigger the hover effect', () => {
      const wrapper = shallow(<PasswordInput onChange={onChangeCallback} />)

      expect(wrapper.prop('focused')).toBe(false)

      wrapper.simulate('mouseenter')
      wrapper.update()

      expect(wrapper.prop('focused')).toBe(true)

      wrapper.simulate('mouseleave')
      wrapper.update()

      expect(wrapper.prop('focused')).toBe(false)
    })
  })
})