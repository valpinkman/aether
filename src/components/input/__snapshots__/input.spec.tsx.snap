// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <EmailInput /> should render 1`] = `
<Styled(styled.div)
  flexDirection="row"
  focused={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  title="wrapper"
  width="400px"
>
  <styled.div
    pr={2}
  >
    <WithSpinner>
      <Styled(User)
        size="20px"
      />
    </WithSpinner>
  </styled.div>
  <styled.input
    bg="transparent"
    color="text.accent"
    fontSize={2}
    maxWidth="100%"
    onChange={[MockFunction]}
    placeholder="email"
    type="email"
    width="100%"
  />
</Styled(styled.div)>
`;

exports[`components <Input /> should render 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": false,
        "rules": Array [
          "
  border: 0;
  outline: none;
  font-weight: bold;
  color: ",
          [Function],
          ";

  ",
          [Function],
          "
  ",
          [Function],
          "
  ",
          [Function],
          "

  ::placeholder {
    color: ",
          [Function],
          ";
  }
",
        ],
      },
      "displayName": "styled.input",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "input",
      "toString": [Function],
      "usesTheme": true,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
/>
`;

exports[`components <PasswordInput /> should render 1`] = `
<Styled(styled.div)
  flexDirection="row"
  focused={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  title="wrapper"
  width="400px"
>
  <styled.div
    pr={2}
  >
    <WithSpinner>
      <Styled(DoorLock)
        size="20px"
      />
    </WithSpinner>
  </styled.div>
  <styled.input
    bg="transparent"
    color="text.accent"
    fontSize={2}
    maxWidth="100%"
    onChange={[MockFunction]}
    placeholder="password"
    type="password"
    width="100%"
  />
  <ShowingBox
    onClick={[Function]}
    opacity={0}
    pl={1}
  >
    <EyeOffIcon
      size="16px"
    />
  </ShowingBox>
</Styled(styled.div)>
`;
